### PR TITLE
[WIP] feat(ui-shell): add side nav components

### LINKED ISSUE #1356: Create Side Navigation component
This issue is for implementing the side navigation component found in:

image

Latest sketch file: https://ibm.box.com/s/4e4zrz6mlo0jpr8lw4vimwlxn4vc59cy

### LINKED ISSUE #1357: Create Global Product Switcher component
This issue is for implementing the header component found in:

image

Latest sketch file: https://ibm.box.com/s/4e4zrz6mlo0jpr8lw4vimwlxn4vc59cy

### LINKED ISSUE #1358: Create Navigation Menu component
This issue is for implementing the navigation menu component found in:

image

Latest sketch file: https://ibm.box.com/s/4e4zrz6mlo0jpr8lw4vimwlxn4vc59cy

### LINKED ISSUE #1359: Create standalone panel components for UI Shell
This issue is for implementing the standalone panel components found in:

image

Latest sketch file: https://ibm.box.com/s/4e4zrz6mlo0jpr8lw4vimwlxn4vc59cy

### PR DESCRIPTION
**Pull request by emyarod**
Closes #1356 

Closes #1357

Closes #1358

Closes #1359 

This PR will implement a side nav and a generic nav panel component, which is extended by a popup nav component as well as a "product switcher" nav base.

#### Changelog

**New**

- side nav component
- navigation menu panel component (generic)
- navigation menu component
- product switcher example component base

**Changed**

- UI shell demo config

### CODE REVIEW

File: src/components/ui-shell/_popup-nav.scss [hunk 1]
```diff
@@ -20,6 +20,11 @@
     color: $ibm-colors__gray--10;
   }
 
+  .#{$prefix}--navigation--right {
+    left: unset;
```
Comment (asudoh) [inline src/components/ui-shell/_popup-nav.scss]: `unset` is not supported by IE. Probably `auto`?

**Review by asudoh — COMMENTED** [review]
LGTM basically 👍, just one thing - Thank you for making this great achievement 🎉 @emyarod!

File: src/components/ui-shell/_ui-shell.scss [hunk 2]
```diff
@@ -4,4 +4,4 @@
 @import 'header';
 @import 'product-switcher';
 @import 'side-nav';
-@import 'nav';
+@import 'popup-nav';
```
Comment (joshblack) [inline src/components/ui-shell/_ui-shell.scss]: For consistency, should we call this navigation menu? Just thinking about this sketch artboard:

image

File: src/components/ui-shell/side-nav.hbs [hunk 3]
```diff
@@ -73,80 +47,18 @@
...
-          </li>
-        </ul>
-      </li>
       <li class="{{@root.prefix}}--side-nav__item {{@root.prefix}}--side-nav__item--active">
-        <button
-          class="{{@root.prefix}}--side-nav__submenu"
-          type="button"
-          aria-haspopup="true"
-          aria-expanded="true">
+        <button class="{{@root.prefix}}--side-nav__submenu" type="button" aria-haspopup="true" aria-expanded="true">
```
Comment (joshblack) [inline src/components/ui-shell/side-nav.hbs]: Any reason for the formatting changes? I ran this block through prettier which is why it's formatted this way.

File: src/components/ui-shell/side-nav.js [hunk 4]
```diff
@@ -1,14 +1,78 @@
...
+    super(element, options);
+    this.manage(on(element, 'click', this._handleClick));
+  }
+
+  /**
+   * Enum for toggling side nav visibility
+   * @readonly
+   * @member SideNav.state
+   * @type {Object}
+   * @property {number} EXPANDED Opening/visible
```
Comment (joshblack) [inline src/components/ui-shell/side-nav.js]: Would this be: 

```suggestion
   * @property {string} EXPANDED Opening/visible
```

**Review by joshblack — CHANGES_REQUESTED** [review]
Left a couple of notes. Looking great!

**Comment by joshblack** [comment]
screen shot 2018-11-16 at 9 40 22 am

Noticed this for the default one 👀  namely the icons in the upper left

Also in upper left:

image

These could easily be handled in the other PR but I wasn't sure 👍 

Also noticed some a11y quirks, namely for whenever we implement a menu widget and the user opens it with enter/spacebar we should place focus on the first item in that menu, I believe.

**Comment by joshblack** [comment]
@emyarod could you add a staging link and add @aagonzales as a reviewer btw?

@aagonzales I'm curious how you feel the focus interactions should work on the side nav 👍

File: src/components/ui-shell/_ui-shell.scss (hunk 2 above)
Comment (emyarod) [reply src/components/ui-shell/_ui-shell.scss]: I renamed the components to `navigation-menu` and `navigation-menu-panel`

**Comment by emyarod** [comment]
the double icon problem should be resolved in https://github.com/IBM/carbon-components/pull/1448/commits/daf5f269793e6921f1a9b47e6505ea8d56dc2eb6

here is the staging link for this component: http://ui-shell-side-nav-busy-squirrel.mybluemix.net/component/ui-shell--default.html

and yes I think we should have a discussion about what we expect in terms of a11y patterns not just for these components but also going forward!

**Comment by joshblack** [comment]
@emyarod the menu pattern specifically is required for ARIA, I believe. Reference comes from: https://www.w3.org/TR/wai-aria-practices-1.1/#menu, namely:

> When a menu opens, or when a menubar receives focus, keyboard focus is placed on the first item. All items are focusable as described in 5.6 Keyboard Navigation Inside Components.

Similarly for the menu button (used in navigation menu) here: https://www.w3.org/TR/wai-aria-practices-1.1/#keyboard-interaction-12

> With focus on the button:
> * Enter: opens the menu and places focus on the first menu item.
> * Space: Opens the menu and places focus on the first menu item.
> * (Optional) Down Arrow: opens the menu and moves focus to the first menu item.
> * (Optional) Up Arrow: opens the menu and moves focus to the last menu item.
> 
> The keyboard behaviors needed after the menu is open are described in 3.14 Menu or Menu bar.

Definitely agreed on the menubar example, we should sync up and document that choice 👍

**Comment by aagonzales** [comment]
> Also noticed some a11y quirks, namely for whenever we implement a menu widget and the user opens it with enter/spacebar we should place focus on the first item in that menu, I believe.

Yes agreed.

How do you navigate down to the sub items? The arrow keys weren't working for me and neither was tab/space.
- oh actually it works with tabs in chrome but I was getting the above issue with firefox

**Review by aagonzales — COMMENTED** [review]
- Can we have all `selected` and `hover` text be white?
- Hover of item in the header should be #4c4c4c
   - Active/click state should be #3D3D3D
- on the right side panel, the overflow menu is hanging of the edge of the screen
image
- overflow menu hover should expand full height of item
   - overflow menu icon also should be centered in container.
image
- There should only be one panel open at a time, so you couldn't have a row of x's. 
image

**Comment by emyarod** [comment]
it seems like either I did not create the staging link for the proper branch or a merge conflict wasn't properly resolved, resulting in broken styles and markup. This doesn't look like what I originally wrote...

I'll try to fix this ASAP

**Comment by emyarod** [comment]
regarding this bug:

> on the right side panel, the overflow menu is hanging of the edge of the screen

this is an issue with overflow menu #1481

**Comment by blazejpawlak** [comment]
Hey, @emyarod any news on this PR?

**Comment by emyarod** [comment]
@blazejpawlak still WIP but I am targeting the next major Carbon release coming up soon

**Comment by emyarod** [comment]
superseded by #1527 #1528 #1529

**Pull request closed without merging**